### PR TITLE
freeipa: init at 4.4.3

### PR DESCRIPTION
**Pull request by outergod**
libverto: init at 0.2.6
pythonPackages.nss: init at 1.0.0
pythonPackages.pki: init at 10.3.5
pythonPackages.lesscpy: init at 0.12.0
pythonPackages.python-yubico: init at 1.3.2

Changed bind to optionally build with Kerberos support.
Changed krb5 to build with system libverto by default.

###### Motivation for this change

This changeset contains all packages and other changes to turn NixOS into a fully capable, well-behaving [FreeIPA](http://www.freeipa.org/page/Main_Page) client and domain member. While only the client portion of FreeIPA is fully implemented, some of the work for building the server component has also been done to actually make the client build succeed.

FreeIPA needs several packages to be build with MIT Kerberos, which in turn needs to use a newer version of libverto than the one bundled with krb5. Because of transitive dependencies, Samba 4 also needs to be build with that same version of MIT Kerberos. The changeset takes a stance that krb5 should much rather always be built with system libverto, which will trigger a mass-rebuild but save time and space in the long run because no duplicate versions of krb5 dependents are needed.
Building with libverto is explicitly switched off for `fetchurl`'s curl derivation to prevent a circular dependency.

The usual way to join a host to IPA is via `freeipa-client-install`, which normally modifies and backs up many different files in `/etc`, which is not possible on NixOS, for obvious reasons. Therefore, I created a nix module that does the same job idiomatically. The only manual steps for a user is to prefetch the certificate from the server and specify the resulting hash in the configuration, as seen below. After activating the configuration, they need to join the domain and obtain the keytab file manually, for which multiple different methods exist, with the most common one presented in the build output.

A typical, a valid set of FreeIPA configuration looks like this:
```
{
  networking.hostName = "nixos.in.foo.com";

  ipa = {
    enable = true;
    domain = "in.foo.com";
    realm  = "IN.FOO.COM";
    server = "ipa.in.foo.com";
    basedn = "dc=in,dc=foo,dc=com";
    certificate = pkgs.fetchurl {
      url = http://ipa.in.foo.com/ipa/config/ca.crt;
      sha256 = "xxx";
    };
  };
  
  services.sssd.enable = true;
  services.ntp.enable = true;
  security.pam.services.login.makeHomeDir = true;  
  security.pam.services.su.makeHomeDir = true;
}
```

The easiest way to test this is to create and set up a FreeIPA instance in a Fedora VM or docker container and build a NixOS version via `nixos-container`, bind-mounting nixpkgs like so:
```
sudo nixos-container create ipa-test
sudo mkdir -p /nix/var/nix/profiles/per-container/ipa-test/per-user/root/channels/nixos
sudo mount --bind /path/to/nixpkgs /nix/var/nix/profiles/per-container/ipa-test/per-user/root/channels/nixos
sudo nixos-container start ipa-test
ln -s /path/to/nixpkgs /path/to/nixpkgs/nixpkgs # Yes, this is required!
sudo nixos-container root-login ipa-test
ipa-test> nix-prefetch-url http://ipa.in.foo.com/ipa/config/ca.crt
ipa-test> cat > /etc/nixos/configuration.nix <<'EOF'
{ config, lib, pkgs, ... }:

with lib;

{ boot.isContainer = true;
  networking.hostName = "ipa-test.in.foo.com";
  networking.useDHCP = false;

  ipa = {
    enable = true;
    domain = "in.foo.com";
    realm  = "IN.FOO.COM";
    server = "ipa.in.foo.com";
    basedn = "dc=in,dc=foo,dc=com";
    certificate = pkgs.fetchurl {
      url = http://ipa.in.foo.com/ipa/config/ca.crt;
      sha256 = "xxx";
    };
  };
  
  services.sssd.enable = true;
  services.ntp.enable = true;
  security.pam.services.login.makeHomeDir = true;  
  security.pam.services.su.makeHomeDir = true;
}
EOF

ipa-test> nixos-rebuild switch
```

I'm well aware this is not a trivial change but I believe the benefit makes up for it and I'm grateful for reviewers who feel like they could tackle this.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (FRidh) [inline pkgs/os-specific/linux/freeipa/default.nix]: please pass in either `python2` or `python2Packages`, not both.

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: while we typically pass in packages individually, in the case of Python we always pass in the whole set. Therefore, pass in either `python2` (and use the pkgs with `python2.pkgs`) or `python2Packages` (and get the interpreter with `python2Packages.python`).

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: this won't be necessary then.

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: could you put this new package, as well as all other new Python packages, in a separate file in `pkgs/development/python-modules/<modulename>/default.nix`.

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: Python packages inherit `meta.platforms` from the interpreter, so no need to pass it here.

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: why are you unsetting this?

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: you could use the new `fetchPypi` function but its not necessary.

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: `makeWrapper` isn't used

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix
Comment (e-user) [inline pkgs/top-level/python-packages.nix]: Please see
https://nixos.org/releases/nixpkgs/nixpkgs-17.03pre91272.7e273d9/manual/#python-setup.py-bdist_wheel-cannot-create-.whl

Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: I'm surprised to see you actually need to unset it yourself.

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: This also requires another name because there is already a `pki` project on PyPi. How about `pki-core`?

**Comment by FRidh** [comment]
tests of `nss` and `pki` are not run

File: pkgs/top-level/python-packages.nix
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: also, is it this package? https://pypi.org/project/python-nss/#files
If so, please give it the name PyPI uses, and preferably fetch from there as well.

**Comment by e-user** [comment]
I've addressed all of the issues you pointed out, @FRidh. `pki-core` doesn't have any tests for Python. All `nss` tests are now being run, but for one of them, an SSL server is started on port 1234, and I'm not sure what NixOS' policies are on tests like that.

**Comment by e-user** [comment]
Ping @FRidh

**Comment by fpletz** [comment]
We have a local freeipa server. I hope I can get around to test this soon. Thanks for your work on this!

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (FRidh) [inline pkgs/os-specific/linux/freeipa/default.nix]: ${python.interpreter}

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: don't add Python libraries here. If its an application, then it should not be called via `python-packages.nix`.

**Review by FRidh — CHANGES_REQUESTED** [review]
This has been open for a quite a while. My apologies for not replying any sooner.

@e-user If there's tests requiring network we disable those.

**Comment by CMCDragonkai** [comment]
Will this lead to allowing NixOS being a freeipa server as well?

**Comment by bjornfor** [comment]
I think not (from first post):

    While only the client portion of FreeIPA is fully implemented, some of the work for building the server component has also been done to actually make the client build succeed.

**Comment by ts468** [comment]
I'd be interested in getting FreeIPA integrated as well!
@e-user: How can I help? It would be a shame to loose all your hard work and effort!
@FRidh, @fpletz: What are your current thoughts and comment about it? Should we pick it up?

File: pkgs/development/python-modules/lesscpy/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/lesscpy/default.nix]: Since recently name in buildPythonPackage can be dropped

File: pkgs/development/python-modules/lesscpy/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/lesscpy/default.nix]: checkInputs

File: pkgs/development/python-modules/nss/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/nss/default.nix]: `${python}`

File: pkgs/development/python-modules/yubico/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/yubico/default.nix]: checkInputs

File: pkgs/development/python-modules/yubico/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/yubico/default.nix]: can be removed

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (FRidh) [inline pkgs/os-specific/linux/freeipa/default.nix]: this can go in the let expression

**Comment by FRidh** [comment]
@ts468 I've only reviewed the Python expressions for which only minor changes are needed. Eventually separate commits are needed as well.

**Comment by Ekleog** [comment]
(triage) @e-user, are you still willing to push this forward? :)

**Comment by e-user** [comment]
Actually, I am interested again, @Ekleog. I'll look into this again over the next days.

**Comment by srhb** [comment]
I, too, am interested here. Anything I can assist with?

**Comment by e-user** [comment]
OK so the push above only takes care of the obvious merge conflicts and @FRidh's requested changes, but I haven't tested the actual implementation after the rebase yet and am going to do so within the next days.

**Comment by e-user** [comment]
OK, so, there's a circular dependency because of curl now depending on libkrb5, which this changeset needs to be built on a proper (as opposed to bundled) version of libverto. Testing will take me a while because of the amounts of rebuilding involved.

**Comment by e-user** [comment]
@FRidh Could you please check whether all of your previous remarks in the review have been addressed?

**Review by FRidh — APPROVED** [review]
Regarding Python this looks good.

**Comment by mmahut** [comment]
What is the status of this pull request?

**Comment by e-user** [comment]
@mmahut Good question. I've been trying to get this one in literally for years now.

**Comment by poelzi** [comment]
We are very much interested in getting this in

**Comment by e-user** [comment]
So yeah, what's left to do? All issues had already been resolved before master moved on, again. I will happily resolve the conflicts, but I would appreciate if that could resolve in an actual merge. Alternatively, if anyone (@poelzi?) would prefer to run a real-life test against an actual server, I'll gladly provide guidance on how to do this, if so desired.

**Comment by mmahut** [comment]
I'm open to test this if you resolve the conflicts.

**Comment by Nepetalactone** [comment]
Apart from the merge conflicts, is there anything that is keeping this from being merged? Would be great to be able to have NixOS machines as FreeIPA clients

**Comment by mmahut** [comment]
ping ^^

**Comment by Nepetalactone** [comment]
@FRidh I hope I am not asking for too much, but since you seem to have merge rights and are already familiar with the pull request, would it be possible that you coordinate the inclusion of this package? There hasn't been any activity on this pull request since half a year and it seems as if there are no missing required changes, apart from the conflicts of course

File: pkgs/development/libraries/libverto/default.nix
Comment (bhipple) [inline pkgs/development/libraries/libverto/default.nix:6]: You can replace this whole thing by just using `pname` below.

File: pkgs/development/libraries/libverto/default.nix
Comment (bhipple) [inline pkgs/development/libraries/libverto/default.nix:34]:
```suggestion
    homepage = "https://github.com/latchset/libverto/";
```

File: pkgs/development/python-modules/nss/default.nix
Comment (bhipple) [inline pkgs/development/python-modules/nss/default.nix:7]:
```suggestion
  pname = "nss";
```

File: pkgs/development/python-modules/nss/default.nix
Comment (bhipple) [inline pkgs/development/python-modules/nss/default.nix:33]:
```suggestion
    homepage    = "https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Python_binding_for_NSS";
```

File: pkgs/development/python-modules/pki-core/default.nix
Comment (bhipple) [inline pkgs/development/python-modules/pki-core/default.nix:5]:
```suggestion
  pname = "pki-core";
```
This makes things much nicer for users pinning this to a different version in an overlay, and matches the general coding standards.

File: pkgs/development/python-modules/pki-core/default.nix
Comment (bhipple) [inline pkgs/development/python-modules/pki-core/default.nix:46]:
```suggestion
    homepage    = "http://pki.fedoraproject.org/";
```

File: pkgs/development/python-modules/yubico/default.nix
Comment (bhipple) [inline pkgs/development/python-modules/yubico/default.nix:21]:
```suggestion
    homepage    = "https://github.com/Yubico/python-yubico";
```

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (bhipple) [inline pkgs/os-specific/linux/freeipa/default.nix:19]:
```suggestion
  pname = "freeipa";
```

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (bhipple) [inline pkgs/os-specific/linux/freeipa/default.nix:102]: Do we have this how, or still the case?

File: pkgs/os-specific/linux/freeipa/default.nix
Comment (bhipple) [inline pkgs/os-specific/linux/freeipa/default.nix:130]:
```suggestion
    homepage = "https://www.freeipa.org/";
```

File: pkgs/os-specific/linux/freeipa/ipa-client-install
Comment (bhipple) [inline pkgs/os-specific/linux/freeipa/ipa-client-install:3]:
```suggestion
echo "ipa-client-install is not available on NixOS. Please see services.ipa.client instead."
```

File: pkgs/servers/dns/bind/default.nix
Comment (bhipple) [inline pkgs/servers/dns/bind/default.nix:6]: Nice :+1:

File: nixos/modules/config/ipa.nix
Comment (bhipple) [inline nixos/modules/config/ipa.nix:140]: Might be nice to make this one configurable.

**Review by bhipple — CHANGES_REQUESTED** [review]
Have a bunch of nit-picky review comments.  Happy to test this further if you rebase/polish/upgrade it.  If no one else objects once that's done I'll merge it in.

File: pkgs/development/compilers/lesscpy/default.nix
Comment (AndersonTorres) [inline pkgs/development/compilers/lesscpy/default.nix:21]:
```suggestion
    homepage    = "https://github.com/lesscpy/lesscpy";
```

File: pkgs/development/libraries/libverto/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/libverto/default.nix:5]:
```suggestion
  pname = "libverto";
  version = "0.2.6";
```

File: pkgs/development/libraries/libverto/default.nix
Comment (AndersonTorres) [inline pkgs/development/libraries/libverto/default.nix:4]:
```suggestion
  pname = "libverto";
```

**Comment by SuperSandro2000** [comment]
Closing since author did not respond in the last months. If this is still important to you please reopen the discussion here.

**Comment by bjornfor** [comment]
Still important to me.

**Comment by SuperSandro2000** [comment]
@bjornfor yeah to you but not to @e-user because he did not respond in 18 months to this. Feel free to pick up the work.

**Comment by e-user** [comment]
Yeah sorry guys. I don't even have a NixOS installation anymore, at all, so I can't really work on it. If anyone needs help finishing this, let me know.

**Comment by benley** [comment]
I'm vaguely considering picking this up.  Before I dive in, does anyone know if there has been other related work that would significantly affect this one?

**Comment by AndersonTorres** [comment]
Almost two years. Just suppose that no one knows.

**Comment by AndersonTorres** [comment]
This is a package database. It will always be interesting to include a package to it.

**Comment by s1341** [comment]
I'm picking this up. I will open a new PR once I get things working...

**Comment by s1341** [comment]
Please see #207115 which supersedes this.

**Pull request closed without merging**
